src/cuda/ptx.cu

```diff
@@ -0,0 +1,262 @@
+#include "cuda_runtime.h"
+#include "device_launch_parameters.h"
+
+#include <stdio.h>
+#include <iostream>
+#include <bitset>
+#include <string>
+
+#include "encode_ptx.h"
+
+#define NUM_THREADS 256
+#define NUM_BLOCKS 1024
+#define EIGHT_GB_IN_BYTES 8589934592
+
+#define CUDA_FATAL(expr) do {				\
+    cudaError_t code = expr;				\
+    if (code != cudaSuccess) {				\
+        cerr << #expr << "@" << __LINE__ << " failed: "	\
+             << cudaGetErrorString(code) << endl;	\
+	exit(1);					\
+    }							\
+} while(0)
+
+
+extern "C" bool detect_cuda()
+{
+    cudaDeviceProp prop;
+    return cudaGetDeviceProperties(&prop, 0) == cudaSuccess;
+}
+
+extern "C" bool batch_encode(unsigned int piece[], size_t len,
+                             const unsigned int iv[32], size_t layers)
+{   // len also represents how many bytes in piece[]
+
+    cudaError_t cudaStatus;  // for handling potential CUDA errors
+
+    int block_count, thread_count;
+    unsigned remaining_piece_size = len;  // there is `size` in the variable name, since len does not represent
+    // the piece count, but instead the size of the piece_array (in bytes)
+    unsigned processed_piece_size = 0;  // in bytes
+
+    thread_count = NUM_THREADS;  // 1 thread is responsible from 1 piece,
+    // 1 thread handles 4096 bytes
+    // 256 threads handle 1048576 bytes, or 2**20 bytes
+
+    unsigned long long default_round_size = EIGHT_GB_IN_BYTES;  // 8GB as Bytes
+    // allocating more than 8GB would be overkill, this is an upper-limit set for high-end GPUs.
+    // we will tweak this down below with respect to the current available device.
+
+    unsigned long long to_be_processed_size, free_mem, total_mem;
+    cudaMemGetInfo(&free_mem, &total_mem);  // getting free and total memory of the device
+
+    printf("\nFree memory on this device is: %llu Bytes\n", free_mem);
```

Comment:
I don't think `printf` is supposed to be in this code. Maybe with some debugging enabled, but definitely not like this.

Reply:
should error message fprintf's be also commented out, until we have a debug config? Or should I keep them uncommented?

Reply:
I would just remove all of these. If you want to know what error happened, return more values than just `0` or `1` and you will be able to match on it and know what went wrong exactly.

Reply:
Sounds good